### PR TITLE
Add support for WAL preallocation

### PR DESCRIPTION
**Pull request by petermattis**
On Linux, preallocation makes a huge difference in sync performance.

```
name                         time/op
SyncPreallocFile/size=0MB-4  503µs ± 6%
SyncPreallocFile/size=4MB-4  303µs ± 2%
SyncReuseFile-4              291µs ± 1%
```

The effect is much more mutated on Darwin, but still present:

```
name                         time/op
SyncPreallocFile/size=0MB-8  50.4µs ±14%
SyncPreallocFile/size=4MB-8  46.4µs ± 2%
SyncReuseFile-8              46.7µs ± 2%
```

Note that the `SyncReuseFile` benchmark is an attempt to measure the win from reusing WAL files. WAL file reuse is not actually implemented  yet.

See #41

### CODE REVIEW

**Comment by petermattis** [comment]
This change is [Reviewable](https://reviewable.io/reviews/petermattis/pebble/76)

**Comment by petermattis** [comment]
Looks like I missed up the perf numbers on Linux as I was measuring `/tmp`. The updated numbers show that preallocation and reuse are fairly close in speed. Here is a more direct view of the delta:

```
name        old time/op  new time/op  delta
SyncFile-4   312µs ± 3%   299µs ± 4%  -4.09%  (p=0.000 n=18+20)
```

"Old" is with preallocation. "New" is with reusing the same file over and over (and no preallocation).

I've verified that I'm definitely using `fdatasync` in both cases.

**Comment by vtolstov** [comment]
Does it possible to unmap from disk some of unneeded wal data? (mostly question for ssd case)

**Comment by petermattis** [comment]
@vtolstov WAL files are currently deleted when they are no longer needed. I'm not sure what you're asking. Can you rephrase your question?

**Comment by vtolstov** [comment]
ah sorry, in this case unmap can happen then file is removed from fs and fs driver code send unmap to this area, sorry.

**Comment by ajkr** [comment]
@vtolstov I think if the filesystem is mounted with discard option, then the unmap/trim command should be sent to the device when the file is deleted. Let me know if I misunderstood your question.

**Comment by vtolstov** [comment]
@ajkr yeah, thanks
